driver-core/src/test/unit/com/mongodb/AuthConnectionStringTest.java

```diff
@@ -133,27 +126,32 @@ private void assertMechanism(final String key, final String actual) {
 
     private void assertMechanismProperties(final MongoCredential credential) {
         BsonValue expected = getExpectedValue("credential.mechanism_properties");
-
-        if (!expected.isNull()) {
-            BsonDocument document = expected.asDocument();
-            for (String key : document.keySet()) {
-                if (document.get(key).isString()) {
-                    String expectedValue = document.getString(key).getValue();
-
-                    // If the mechanism is "GSSAPI", the default SERVICE_NAME, which is stated as "mongodb" in the specification,
-                    // is set to null in the driver.
-                    if (credential.getMechanism().equals("GSSAPI") && key.equals("SERVICE_NAME") && expectedValue.equals("mongodb")) {
-                        assertNull(credential.getMechanismProperty(key, null));
-                    } else {
-                        assertEquals(expectedValue, credential.getMechanismProperty(key, null));
-                    }
+        if (expected.isNull()) {
+            return;
+        }
+        BsonDocument document = expected.asDocument();
+        for (String key : document.keySet()) {
+            Object actualMechanismProperty = credential.getMechanismProperty(key, null);
+            if (document.get(key).isString()) {
+                String expectedValue = document.getString(key).getValue();
+                // If the mechanism is "GSSAPI", the default SERVICE_NAME, which is stated as "mongodb" in the specification,
+                // is set to null in the driver.
+                if (credential.getMechanism().equals("GSSAPI") && key.equals("SERVICE_NAME") && expectedValue.equals("mongodb")) {
+                    assertNull(actualMechanismProperty);
                 } else {
-                    assertEquals(document.getBoolean(key).getValue(), credential.getMechanismProperty(key, (Boolean) null).booleanValue());
+                    assertEquals(expectedValue, actualMechanismProperty);
                 }
+            } else if ((document.get(key).isBoolean())) {
+                boolean expectedValue = document.getBoolean(key).getValue();
+                assertNotNull(actualMechanismProperty);
+                assertEquals(expectedValue, actualMechanismProperty);
+            } else {
+                fail("unsupported property type");
             }
         }
     }
 
+    @Nullable
```

Comment:
`getExpectedValue` was annotated with `@Nullable`, yet it is always used as if it were not able to return `null`.